rfcs/0048-rosetta-zkapps.md

```diff
@@ -0,0 +1,119 @@
+
+#### Construction API
+
+As such, custom tokens are _out of scope_ and thus the Rosetta Construction flow will _not_ need to change. Instead, we will continue to support the existing legacy transactions.
+
+#### Data API
+
+The [Data API](https://www.rosetta-api.org/docs/data_api_introduction.html) will need to change to support zkApps.
+
+Most of the changes will be localized to supporting new zkApps related [Rosetta operations](https://www.rosetta-api.org/docs/models/Operation.html) and putting them in a new type of [Rosetta transaction](https://www.rosetta-api.org/docs/models/Transaction.html).
+
+These new transactions will be present in both the `/block` and `/mempool` endpoints.
+
+Note: GraphQL queries for zkApp transactions can be [generated programatically](https://github.com/MinaProtocol/mina/blob/develop/src/lib/mina_base/parties.ml#L1431) as they are quite large.
+
+**Operations for zkApps**
+
+[Operation types](https://github.com/MinaProtocol/mina/blob/35ed5e191af9cfa2709f567f6fe85d96dabfafef/src/lib/rosetta_lib/operation_types.ml) should be extended with new operations for all kinds of ways zkApp transactions can manipulate account balances (for each one-sided changes).
+
+Luckily this is made extremely clear by the shape of the zkApps transaction structure.
+
+The following intends to be an exhaustive list of those sorts of operations:
+
+1. `Zkapps_fee_payer_dec`
+2. `Zkapps_balance_change`
+
+Note that only balance changes that correspond with the default MINA token (`token_id` = 1) should be considered as changes here.
+
+There are many new types of failures and they're per-party in zkApps transactions. These are enumerated in [`transaction_status.ml`](https://github.com/MinaProtocol/mina/blob/a6e5f182855b3f4b4afb0ea8636760e618e2f7a0/src/lib/mina_base/transaction_status.ml). They can be pulled from the `zkapp_party_failures` table in the archive database. Since the errors are already per-party, they're easily associated with operations on a one-to-one basis. These can be added verbatim to as ["reason metadata"](https://github.com/MinaProtocol/mina/blob/a6e5f182855b3f4b4afb0ea8636760e618e2f7a0/src/lib/rosetta_lib/user_command_info.ml#L449) to those operations with a `Failed` label.
+
+### Supporting hardforks
+
+Changes will need to be made to `/network/status` such that the genesis block is the first block of the hardfork network (so it's index will be greater than 1; to be determined when we fork). This is important to be the true block height or else the archive data queries will be incorrect.
+
+`/network/list` will remain unchanged as in the hardfork world the old network "doesn't exist".
+
+### Testing
+
+#### Rosetta CLI
+
+Rosetta has a [cli testing tool](https://github.com/coinbase/rosetta-cli) that must pass on this network.
```

Comment:
should we mention the migration of the old mainnet data as part of the testing?

Reply:
Hmm do we need the old data? I thought we're recreating the Genesis ledger after the hardfork? I don't recall what we decided here?

Reply:
Don't we need the old data for historical lookups? Is that no longer considered part of the chain?

Reply:
I think that is no longer considered part of the chain (there is an option in Rosetta to start the chain at a non-zero block height)